### PR TITLE
[#24789][prism] V0 Go Direct Runner Replacement, Prism.

### PR DESCRIPTION
**Pull request by lostluck**
Add the final connections for making the runner usable.
Add in the pipeline execution tests, and initial separation harness tests.

Note, SDF splitting isn't yet implemented, although non-splitting expansion is supported.

See #24789. This doesn't complete the the issue as the issue is intended to replace the default Go SDK runner. This PR doesn't yet set prism as default, which would require at minimum instructions for switching back to the legacy direct runner, and for repository infrastructure to test on presubmit.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by lostluck** [comment]
R: @johannaojeling 
cc: @jrmccluskey 

As foretold, this is a big one, but it's the last one in this veritable barrage of PRs. Thank you very much!

File: sdks/go/pkg/beam/runners/prism/internal/separate_test.go [hunk 1]
```diff
@@ -0,0 +1,593 @@
...
+	"github.com/apache/beam/sdks/v2/go/pkg/beam/core/sdf"
+	"github.com/apache/beam/sdks/v2/go/pkg/beam/core/util/reflectx"
+	"github.com/apache/beam/sdks/v2/go/pkg/beam/io/rtrackers/offsetrange"
+	"github.com/apache/beam/sdks/v2/go/pkg/beam/register"
+	"github.com/apache/beam/sdks/v2/go/pkg/beam/testing/passert"
+	"github.com/apache/beam/sdks/v2/go/pkg/beam/transforms/stats"
+	"golang.org/x/exp/slog"
+)
+
+// separate_test.go is retains structures and tests to ensure the runner can
```
Comment (johannaojeling) [inline sdks/go/pkg/beam/runners/prism/internal/separate_test.go:40]:
```suggestion
// separate_test.go retains structures and tests to ensure the runner can
```

File: sdks/go/pkg/beam/runners/prism/internal/separate_test.go [hunk 2]
```diff
@@ -0,0 +1,593 @@
...
+// Global variable, so only one is registered with the OS.
+var ws = &Watchers{}
+
+// TestSeparation validates that the runner is able to split
+// elements in time and space. Beam has a few mechanisms to
+// do this.
+//
+// First is channel splits, where a slowly processing
+// bundle might have it's remaining buffered elements truncated
+// so they can be processed by a another bundle,
```
Comment (johannaojeling) [inline sdks/go/pkg/beam/runners/prism/internal/separate_test.go:52]:
```suggestion
// so they can be processed by another bundle,
```

File: sdks/go/pkg/beam/runners/prism/internal/separate_test.go [hunk 3]
```diff
@@ -0,0 +1,593 @@
...
+// Watchers is a "net/rpc" service.
+type Watchers struct {
+	mu             sync.Mutex
+	nextID         int
+	lookup         map[int]*watcher
+	serviceOnce    sync.Once
+	serviceAddress string
+}
+
+// Args is the set of parameters to the watchers RPC methdos.
```
Comment (johannaojeling) [inline sdks/go/pkg/beam/runners/prism/internal/separate_test.go:192]:
```suggestion
// Args is the set of parameters to the watchers RPC methods.
```

File: sdks/go/pkg/beam/runners/prism/internal/separate_test.go [hunk 4]
```diff
@@ -0,0 +1,593 @@
...
+		return fmt.Errorf("no watcher with id %v", args.WatcherID)
+	}
+	w.mu.Lock()
+	*unblocked = w.sentinelCount >= w.sentinelCap
+	w.mu.Unlock()
+	slog.Debug("sentinel target for watcher%d is %d/%d. unblocked=%v", args.WatcherID, w.sentinelCount, w.sentinelCap, *unblocked)
+	return nil
+}
+
+// Delay returns whether the sentinels shoudld delay.
```
Comment (johannaojeling) [inline sdks/go/pkg/beam/runners/prism/internal/separate_test.go:227]:
```suggestion
// Delay returns whether the sentinels should delay.
```

File: sdks/go/pkg/beam/runners/prism/internal/separate_test.go [hunk 5]
```diff
@@ -0,0 +1,593 @@
...
+			panic(err)
+		}
+		rpc.Register(ws)
+		rpc.HandleHTTP()
+		go http.Serve(l, nil)
+		ws.serviceAddress = l.Addr().String()
+	})
+}
+
+// newWatcher starts an rpc server to maange state for watching for
```
Comment (johannaojeling) [inline sdks/go/pkg/beam/runners/prism/internal/separate_test.go:259]:
```suggestion
// newWatcher starts an rpc server to manage state for watching for
```

File: sdks/go/pkg/beam/runners/prism/internal/separate_test.go [hunk 6]
```diff
@@ -0,0 +1,593 @@
...
+		client, err := rpc.DialHTTP("tcp", fn.LocalService)
+		if err != nil {
+			slog.Error("failed to dial sentinels  server", err, slog.String("endpoint", fn.LocalService))
+			panic(fmt.Sprintf("dialing sentinels server %v: %v", fn.LocalService, err))
+		}
+		sepClient = client
+		sepWaitMap = map[int]chan struct{}{}
+	})
+
+	// Check if there's alreaedy a local channel for this id, and if not
```
Comment (johannaojeling) [inline sdks/go/pkg/beam/runners/prism/internal/separate_test.go:304]:
```suggestion
	// Check if there's already a local channel for this id, and if not
```

File: sdks/go/pkg/beam/runners/prism/internal/separate_test.go [hunk 7]
```diff
@@ -0,0 +1,593 @@
...
+	// start a watcher goroutine to poll and unblock the harness when
+	// the expected number of ssentinels is reached.
```
Comment (johannaojeling) [inline sdks/go/pkg/beam/runners/prism/internal/separate_test.go:306]:
```suggestion
	// the expected number of sentinels is reached.
```

**Review by johannaojeling — COMMENTED** [review]
These may be my personal preferences, but I think the code would be easier to follow if `RunPipeline`, as the main function in the internal package, was defined in the beginning of execute.go, followed by the functions it uses (`runEnvironment`, `executePipeline`, and so on).

It might also make sense to move `stage` from execute.go to preprocess.go, or to its own file if you would prefer that. As it is now, `executePipeline()` in execute.go initializes `preprocessor` and invokes `preprocessor.preProcessGraph()`, which is defined in preprocess.go. Whereas preprocessor.preProcessGraph() initializes stage, which is defined in execute.go.

Which is fine for now from a compilation perspective since they are all in the same package, but from a design perspective these dependencies are messing with my mind a bit, thinking that if they were in separate packages we would have an import cycle.

Other than that LGTM. Congrats on finalizing this last piece of the initial version!

**Comment by lostluck** [comment]
Ready for another look!

> These may be my personal preferences, but I think the code would be easier to follow if `RunPipeline`, as the main function in the internal package, was defined in the beginning of execute.go, followed by the functions it uses (`runEnvironment`, `executePipeline`, and so on).
> 

Fair! I re-arranged a little before sending this out. That's the reverse of my usual preference (having things defined before they're used). But I agree that ordering is more approachable for reading how the runner works in this case. 

> It might also make sense to move `stage` from execute.go to preprocess.go, or to its own file if you would prefer that. As it is now, `executePipeline()` in execute.go initializes `preprocessor` and invokes `preprocessor.preProcessGraph()`, which is defined in preprocess.go. Whereas preprocessor.preProcessGraph() initializes stage, which is defined in execute.go.

This one is trickier. Preprocess.go only uses a single field in stage: the slice of transform IDs, everything else is used around this file, and is more central to execution. So it shouldn't live in preprocess.go IMO.

Preprocess used to live in execute.go but the preprocessor, while simple now, will eventually handle stage fusion, and become more complicated, so it's worth splitting to it's own file for testing. A bunch of stuff should probably moved to the pre-processor too.

But I've ended up moving it to its own file in the end, because it turns out it's most of the content? Some of it will need to become its own "handler", for handling messages to the FnAPI. I had clearly lost scope about how much code was unique to building and managing stages. Ultimately, it'll be a candidate to move to it's own package along with associated unit testing and documentation.

**Comment by johannaojeling** [comment]
Makes sense and looks good 🙂

**Pull request merged**